plugins/chat/assets/javascripts/lib/rich-editor-extension.js

```diff
@@ -0,0 +1,245 @@
+      // The slice here makes sure we get the html_raw content
+      // only for the current [chat] bbcode block based on the
+      // token index.
+      const nextHtmlRaw = tokens
+        .slice(i)
+        .find((t) => t.type === "html_raw" && t.nesting === 1);
+      const messagesHtml = nextHtmlRaw?.content;
+
+      let threadHtmlRaw;
+      if (token.attrGet("data-thread-id")) {
+        threadHtmlRaw = tokens
+          .slice(tokens.indexOf(nextHtmlRaw) + 1)
+          .find((t) => t.type === "html_raw" && t.nesting === 1);
+      }
+
+      // So this content and the whole wrap_open happens for every single instance of
+      // the `[chat]` bbcode, including nested which is the case for threads.
+      //
+      // Only the first one will have multiQuote and chained
+      //
+      // Multiquote is > 1 message
+      // Chained is > 1 message by different users
+      state.openNode(state.schema.nodes.chat, {
+        messageId: token.attrGet("data-message-id"),
+        username: token.attrGet("data-username"),
+        datetime: token.attrGet("data-datetime"),
+        channelName: token.attrGet("data-channel-name"),
+        channelId: token.attrGet("data-channel-id"),
+        chained: token.attrGet("data-chained"),
+        multiQuote: token.attrGet("data-multiquote"),
+        threadId: token.attrGet("data-thread-id"),
+        threadTitle: token.attrGet("data-thread-title"),
+        threadHtml: threadHtmlRaw ? threadHtmlRaw.content : null,
+        html: messagesHtml,
+        rawContent: token.content,
+      });
+      return true;
+    },
+    div_chat_transcript_wrap_close(state) {
+      state.closeNode();
+      return true;
+    },
+    div_chat_transcript_user: { ignore: true },
+    div_chat_transcript_user_avatar: { ignore: true },
+    div_chat_transcript_username: { ignore: true },
+    div_chat_transcript_datetime: { ignore: true },
+    div_chat_transcript_messages: { ignore: true },
+    div_chat_transcript_reaction: { ignore: true },
+
+    // Reaction-related tokens are not used in the live preview,
+    // they are only used when archiving a channel, not needed here.
+    div_chat_transcript_reactions: { ignore: true },
+    div_chat_transcript_meta: { ignore: true },
+
+    // Thread-related tokens
+    details_chat_transcript_wrap: { ignore: true },
+    summary_chat_transcript: { ignore: true },
+    div_thread: { ignore: true },
+    div_thread_header: { ignore: true },
+    svg_thread_header: { ignore: true },
+    use_svg_thread: { ignore: true },
+    span_thread_title: { ignore: true },
+
+    html_raw: { ignore: true, noCloseToken: true },
+
+    span: { ignore: true },
+    span_open() {
+      // TODO: not sure if i need to actually do anything here,
+      // its just here to stop it erroring, otherwise I get
+      //
+      // No parser processed span_open token for tag: span, attrs: [["title","2025-03-20T07:12:57Z"]]
+      return true;
```

Comment:
@renato no idea what to do here and the error message is cryptic

Reply:
Hmm seems to be fine without this now, weird. If I do include it local dates breaks

Reply:
Oh, it's because there's still a `span_open` token that the parser needs to process.. if you `return true` you're saying you're handling all occurrences of `span_open`, which is a token type that is being re-used in different places for different purposes. It's not _clean_, but we may remove it from the token stream.

Reply:
https://github.com/discourse/discourse/blob/840b3d67e12ca8df4788558f8051612eaeea4489/plugins/chat/assets/javascripts/lib/discourse-markdown/chat-transcript.js#L215-L248

The outcome of this logic is that we have some `span_open` tokens without content

![image](https://github.com/user-attachments/assets/fb97d255-b136-4a7c-8436-a73960dd3b70)

The content is only added here:

https://github.com/discourse/discourse/blob/e591d89e73a7845061e617985550560f294211a2/plugins/chat/assets/javascripts/discourse/initializers/chat-setup.js#L121-L147

(You probably already knew this, but I'm discovering this now 😄)

Reply:
Nope I didn't know this, thanks for explaining :)